### PR TITLE
Refactored Tree Graph Final Hope

### PR DESCRIPTION
**Pull request by tharmaman**

### CODE REVIEW

File: src/foam/graphics/TreeGraph.js [hunk 1]
```diff
@@ -33,45 +33,26 @@
      {
        name: 'formatNode',
        value: function() {}
-     }
+     },
+     'time'
```
Comment (mcarcaso) [inline src/foam/graphics/TreeGraph.js]: Can we remove this now?

File: src/foam/graphics/TreeGraph.js [hunk 2]
```diff
@@ -33,45 +33,26 @@
...
+   ],
+
+   topics: [
+    'layoutComplete'
```
Comment (mcarcaso) [inline src/foam/graphics/TreeGraph.js]: Let's call this 'resizeComplete'

File: src/foam/graphics/TreeGraph.js [hunk 3]
```diff
@@ -302,37 +283,53 @@
...
-          maxRight: 0
-        };
-          
-        function traverseAndCompare(root) {
-          if ( root.maxLeft < maxes.maxLeft ) maxes.maxLeft = root.maxLeft;
-          if ( root.maxRight > maxes.maxRight ) maxes.maxRight = root.maxRight;
-
-          for ( var i = 0; i < root.children.length; i++ ){
-            traverseAndCompare(root.children[i]);
+         if ( ! this.time ) {
```
Comment (mcarcaso) [inline src/foam/graphics/TreeGraph.js]: Remove the timing

File: src/foam/graphics/TreeGraph.js [hunk 4]
```diff
@@ -302,37 +283,53 @@
...
+           this.time = Date.now();
+         }
+          if ( this.root && this.root.layout() ) {
+            this.invalidate();
+            this.doLayout();
+          } else {
+            // console.log(Date.now() - this.time)
```
Comment (mcarcaso) [inline src/foam/graphics/TreeGraph.js]: REmove

File: src/foam/graphics/TreeGraph.js [hunk 5]
```diff
@@ -302,37 +283,53 @@
...
+         })
+
+         var width = Math.abs(maxes.maxLeft - maxes.maxRight);
+         var height = (this.nodeHeight * 2) * (this.root.outline.length + 1);
+
+         var deltaX = Math.abs(this.width - width) / width;
+         var deltaY = Math.abs(this.height - height) / height;
+
+         if ( deltaX > 0.01 || deltaY > 0.01 ) {
+          //  debugger;
```
Comment (mcarcaso) [inline src/foam/graphics/TreeGraph.js]: Remove

File: src/foam/graphics/TreeGraph.js (hunk 1 above)
Comment (tharmaman) [reply src/foam/graphics/TreeGraph.js]: Ye

File: src/foam/graphics/TreeGraph.js (hunk 2 above)
Comment (tharmaman) [reply src/foam/graphics/TreeGraph.js]: Yer

File: src/foam/graphics/TreeGraph.js (hunk 3 above)
Comment (tharmaman) [reply src/foam/graphics/TreeGraph.js]: Yea

File: src/foam/graphics/TreeGraph.js (hunk 4 above)
Comment (tharmaman) [reply src/foam/graphics/TreeGraph.js]: Ya

File: src/foam/graphics/TreeGraph.js (hunk 5 above)
Comment (tharmaman) [reply src/foam/graphics/TreeGraph.js]: Yes

**Pull request closed without merging**